// @vsintellicode/typescript-intellicode-plugin/index.js
/*! Copyright (c) Microsoft Corporation. All rights reserved. */
module.exports=function(e){var n={};function t(o){if(n[o])return n[o].exports;var i=n[o]={i:o,l:!1,exports:{}};return e[o].call(i.exports,i,i.exports,t),i.l=!0,i.exports}return t.m=e,t.c=n,t.d=function(e,n,o){t.o(e,n)||Object.defineProperty(e,n,{enumerable:!0,get:o})},t.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},t.t=function(e,n){if(1&n&&(e=t(e)),8&n)return e;if(4&n&&"object"==typeof e&&e&&e.__esModule)return e;var o=Object.create(null);if(t.r(o),Object.defineProperty(o,"default",{enumerable:!0,value:e}),2&n&&"string"!=typeof e)for(var i in e)t.d(o,i,function(n){return e[n]}.bind(null,i));return o},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,"a",n),n},t.o=function(e,n){return Object.prototype.hasOwnProperty.call(e,n)},t.p="",t(t.s=0)}([function(e,n,t){"use strict";var o=this&&this.__importStar||function(e){if(e&&e.__esModule)return e;var n={};if(null!=e)for(var t in e)Object.hasOwnProperty.call(e,t)&&(n[t]=e[t]);return n.default=e,n};const i=o(t(1)),r=o(t(3)),s=o(t(4)),a=o(t(5)),c=o(t(6)),u=1e6,l=1e9;let f,d,p=!1,m=500,y=void 0;function g(e,n,t){const o=S(...process.hrtime(e));n&&"None"===n.failureReason?t.metadata={type:n.foundNode.typeName,inIf:n.foundNode.isInConditional,patterns:n.sequences,computationTimeMS:o,failureReason:"None"}:t.metadata={computationTimeMS:o,failureReason:n?n.failureReason:"LookupKeyGenerationTimeout"}}function h(e,n){e.languageServiceHost.log&&e.languageServiceHost.log(`IntelliCode plugin: ${n}`)}function S(e,n){return(e*l+n)/u}const N=3.4,x=3.6,T=3.7;class k{constructor(e,n,t){this.ts=e,this.hostCancellationToken=n,this.timeoutMS=t,this.startTime=Date.now()}throwIfCancellationRequested(){if(this.isHostTokenCanceled())throw new v(!0);if(p&&this.isLocalTokenCanceled())throw new v(!1)}isHostTokenCanceled(){return this.hostCancellationToken.isCancellationRequested()}isLocalTokenCanceled(){return Math.abs(Date.now()-this.startTime)>=this.timeoutMS}}class v{constructor(e){this.canceledByHostToken=e}}e.exports=function(e){const n=e.typescript;let t;return{create:function(e){if(!n.versionMajorMinor)return e.languageService;const u=Number(n.versionMajorMinor);if(NaN===u||u<3.2)return e.languageService;const l=Object.create(null);for(let n of Object.keys(e.languageService)){const t=e.languageService[n];t&&(l[n]=t.bind(e.languageService))}function C(){if(!y&&f)try{const n=s.readFileSync(f),t=Buffer.from([214,105,172,23,156,35,36,45,137,37,195,176,189,110,127,78,69,171,205,160,3,202,114,220,252,109,233,221,252,181,45,47]),o=Buffer.from([60,12,128,217,163,220,176,146,213,18,190,59,212,88,192,21]),i=r.createDecipheriv("aes-256-cbc",t,o),c=Buffer.concat([i.update(n),i.final()]),u=a.gunzipSync(c);y=JSON.parse(u.toString("utf-8")),d="typescript_"+(y.Version||"unknown"),h(e,`successfully loaded model from ${f}`)}catch(n){h(e,`failed to load model from ${f}, exception: ${n}`)}return y}return o(e.config),t=e.languageServiceHost,h(e,`starting configuration: ${JSON.stringify(e.config)}`),l.getCompletionsAtPosition=(t,o,r)=>(function(e,n,t,o,r,s,a){const u=t.languageService.getCompletionsAtPosition(r,s,a);if(!u||!u.isMemberCompletion)return u;const l=t.project.getCancellationToken(),f=new k(e,l,m),y=process.hrtime(),C=t.languageService.getProgram(),b=C?C.getSourceFile(r):void 0;if(u.entries.length>0||!b||!c.isInString(e,b,s)){if(p){try{f.throwIfCancellationRequested();const e=O(f);g(y,e,u)}catch(n){if(!(n instanceof v))throw n;if(n.canceledByHostToken)throw new e.OperationCanceledException;g(y,void 0,u)}return u}const i=function(e){let n=void 0,i=[];const r=o(),s=r&&r.model;if(!s)return{recommendedWords:[],failureReason:"NoModelAvailable"};const a=O(e);if(!a.foundNode||!a.sequences)return{recommendedWords:[],failureReason:a.failureReason};const c=s[a.foundNode.typeName];if(c){h(t,"found some suggestion stuff");for(const o of a.sequences){e.throwIfCancellationRequested();const r=c[o],s=r&&(a.foundNode.isInConditional?r[1]:r[0]);if(s){n=o,h(t,`found suggestions for sequence ${o} : ${s.join(",")}`),i.push(...s.slice(0,5));break}}}return{typeName:a.foundNode.typeName,inConditional:a.foundNode.isInConditional,sequenceString:n,recommendedWords:i,failureReason:"None"}}(f),r=S(...process.hrtime(y));let s="None"===i.failureReason?"NotInModel":i.failureReason,a=[];i.recommendedWords.length>0&&(s="NotInIntersection",u.entries=u.entries.filter(e=>!e.replacementSpan||!e.insertText||"["!==e.insertText[0]),u.entries.some(e=>(function(e,n){const t=n.sortText;if(e<=N)return!t.endsWith("1");if(e<=x)return!t.endsWith("4");if(e<=T)return!t.endsWith("6");return!n.isFromUncheckedFile})(n,e))?i.recommendedWords.forEach((e,n)=>{const o=u.entries.findIndex(n=>n.name===e);if(o>=0){const e=u.entries[o];h(t,`marking recommendation ${e.name}`),s="None",I(e,n),a.push(e),u.entries.splice(o,1)}}):(h(t,"no strong recommendations, adding entries"),u.entries=u.entries.filter(e=>!i.recommendedWords.some(n=>e.name===n)),a=i.recommendedWords.map((n,t)=>{const o={name:n,kind:e.ScriptElementKind.memberVariableElement,sortText:"0"};return I(o,t),o})),u.entries=a.concat(u.entries));const c=function(e,n,t){return{Id:Date.now(),ModelVersion:d||"typescript_unknown",Count:0,FailureReason:e,Class:"",InIf:n.inConditional,ElapsedTime:t,Methods:[],PrevInvoc:"",ItemMetadata:Object.create(null)}}(s,i,r);i.typeName&&i.sequenceString&&(c.Count=a.length,c.ModelType=i.sequenceString&&"N"!==i.sequenceString?"Sequence":"Frequency",c.Class="NotInModel"!==s?i.typeName:"",c.Methods=a.map(e=>e.insertText),c.PrevInvoc=i.sequenceString,u.entries.forEach((e,n)=>{const o={Id:c.Id,Index:n,Method:n<a.length?e.insertText:""};c.ItemMetadata[e.name]=o,h(t,`adding item metadata ${e.name} => ${JSON.stringify(o)} to completion result`)})),h(t,`added metadata ${JSON.stringify(c)} to completion result`),u.metadata=c}function I(e,n){e.insertText=e.insertText||e.name,e.sortText=e.sortText||e.name,e.name="★ "+e.name,e.sortText=`*${String.fromCharCode("a".charCodeAt(0)+n)} ${e.sortText}`,h(t,`sort text for entry ${e.name} is ${e.sortText}`)}function O(n){const o=t.languageService.getProgram();if(!o)return{failureReason:"NoProgramFound"};h(t,"program found"),n.throwIfCancellationRequested();const a=o.getTypeChecker(),u=o.getSourceFile(r);if(!u)return{failureReason:"NoSourceFileFound"};h(t,"source file found"),n.throwIfCancellationRequested();const l=function(n,t,o){let r=c.findPrecedingToken(e,o,t);r&&o<=r.end&&(e.isIdentifier(r)||c.isKeyword(e,r.kind))&&(r=c.findPrecedingToken(e,r.getFullStart(),t));if(!r||r.kind!==e.SyntaxKind.DotToken)return;const s=r.parent;if(!e.isPropertyAccessExpression(s))return;const a=i.getTypeNameForPythia(e,i.GetTypeNameOptions.SkipAny,n,s.expression,t).typeName;if(!a)return;return{isInConditional:A(s),node:s,typeName:a,sf:t}}(a,u,s);if(n.throwIfCancellationRequested(),!l)return{failureReason:"NoNodeFound"};h(t,`determined type: ${l.typeName}`);const f=function(o){const r=[];try{!function o(r,s,c,u){if(s.getStart()>c.getStart()||u.length>=2)return;n.throwIfCancellationRequested();const f=s.getChildren(l.sf);for(let e=f.length-1;e>=0;e--)o(r,f[e],c,u);if(e.isPropertyAccessExpression(s)&&s!=c){i.getTypeNameForPythia(e,i.GetTypeNameOptions.SkipAny,a,s.expression,l.sf).typeName===r&&s.name.text&&(h(t,`pushing prior access: ${s.name.text}`),u.push(s.name.text))}}(o,l.sf,l.node,r)}catch(n){if(!(n instanceof v))throw n;if(n.canceledByHostToken)throw new e.OperationCanceledException}h(t,`prior accesses: ${r.join(",")}`);const s=[];r.length>1&&s.push(r[1]+"~"+r[0]);r.length>0&&s.push("N~"+r[0]);return s.push("N"),h(t,`sequences: ${s.join(",")}`),s}(l.typeName);return h(t,`sequences: ${f.join(",")}`),{foundNode:l,sequences:f,failureReason:"None"}}function A(n){let t=n;for(;t.kind!==e.SyntaxKind.SourceFile;){const n=t.parent;if(e.isIfStatement(n)&&n.expression===t||e.isConditionalExpression(n)&&n.condition===t||e.isDoStatement(n)&&n.expression===t||e.isWhileStatement(n)&&n.expression===t||e.isForStatement(n)&&n.condition===t)return!0;t=n}return!1}return u})(n,u,e,C,t,o,r),l.getCompletionEntryDetails=(...t)=>(function(e,n,t,o,i,r,...s){const a=r.startsWith("★ "),c=a?r.slice("★ ".length):r;a&&h(n,`stripped name for recommendation: ${c}`);const u=(0,n.languageService.getCompletionEntryDetails)(o,i,c,...s);u&&a&&(u.name="★ "+u.name,h(n,`starred name for recommendation: ${u.name}`));return u})(n,e,C,...t),l},onConfigurationChanged:function(e){t&&t.log(`received configuration command: ${JSON.stringify(e)}`),o(e)}};function o(e){e.modelPath&&(f=e.modelPath),e.returnLookupKey&&(p=e.returnLookupKey),e.timeout&&(m=e.timeout)}}},function(e,n,t){"use strict";n.__esModule=!0;var o,i,r=t(2);!function(e){e[e.DontSkipAny=0]="DontSkipAny",e[e.SkipAny=1]="SkipAny"}(o=n.GetTypeNameOptions||(n.GetTypeNameOptions={})),function(e){e[e.SkipAny=0]="SkipAny",e[e.Any=1]="Any",e[e.AnyCallExpression=2]="AnyCallExpression",e[e.AnyPropertyAccessExpression=3]="AnyPropertyAccessExpression",e[e.AnyElementAccessExpression=4]="AnyElementAccessExpression",e[e.UnionOrIntersection=5]="UnionOrIntersection",e[e.OtherPrimitive=6]="OtherPrimitive",e[e.OtherNonPrimitive=7]="OtherNonPrimitive",e[e.Literal=8]="Literal",e[e.EmptyIdentifier=9]="EmptyIdentifier"}(i=n.SkipTypeReason||(n.SkipTypeReason={})),n.getTypeNameForPythia=function e(n,t,s,a,c){var u=s.getTypeAtLocation(a),l=u.aliasSymbol||u.getSymbol();if(l){if(0!=(l.flags&(n.SymbolFlags.TypeLiteral|n.SymbolFlags.ObjectLiteral)))return{typeName:s.typeToString(u)};var f=s.getFullyQualifiedName(l);if(f&&'"'===f[0]){var d=function(e){var n=e.indexOf('"',1);if(n>=1)return{partInQuotes:e.substring(1,n),rest:e.substring(n+1)};return{}}(f),p=d.partInQuotes,m=d.rest;if(p){var y=p;if(r.isAbsolute(y)){"index"===r.basename(y)&&(y=r.dirname(y));var g=y.lastIndexOf("/node_modules/");if(g>=0){var h=g+"/node_modules/".length;y=y.substring(h)}else y=r.basename(y);f='"'+y+'"'+(m||"")}}}return{typeName:f}}if(0==(u.getFlags()&n.TypeFlags.NonPrimitive)){if(0!=(u.flags&n.TypeFlags.Any)){var S;if(n.isCallExpression(a))return(S=e(n,t,s,a.expression,c)).typeName?{typeName:S.typeName+"()"}:{skipReason:i.AnyCallExpression};if(n.isPropertyAccessExpression(a))return(S=e(n,t,s,a.expression,c)).typeName?{typeName:S.typeName+"."+a.name.text}:{skipReason:i.AnyPropertyAccessExpression};if(n.isElementAccessExpression(a))return(S=e(n,t,s,a.expression,c)).typeName?{typeName:S.typeName+"[]"}:{skipReason:i.AnyElementAccessExpression};if(t===o.SkipAny)return{skipReason:i.SkipAny};if(n.isIdentifier(a)){var N=a.text;return N?{typeName:N}:{skipReason:i.EmptyIdentifier}}return a.kind===n.SyntaxKind.ThisKeyword?{typeName:"this"}:{skipReason:i.Any}}return 0!=(u.flags&n.TypeFlags.StringLike)?{typeName:"string"}:0!=(u.flags&n.TypeFlags.Boolean)?{typeName:"boolean"}:0!=(u.flags&n.TypeFlags.NumberLike)?{typeName:"number"}:0!=(u.flags&n.TypeFlags.UnionOrIntersection)?{skipReason:i.UnionOrIntersection}:{skipReason:i.OtherPrimitive}}return{skipReason:i.OtherNonPrimitive}}},function(e,n){e.exports=require("path")},function(e,n){e.exports=require("crypto")},function(e,n){e.exports=require("fs")},function(e,n){e.exports=require("zlib")},function(e,n,t){"use strict";function o(e,n,t){return n.kind===e.SyntaxKind.EndOfFileToken?!!n.jsDoc:0!==n.getWidth(t)}function i(e,n){return e.isJsxText(n)&&(n.containsOnlyTriviaWhiteSpaces||n.containsOnlyWhiteSpaces)}function r(e,n,t,r){for(let s=t-1;s>=0;s--){if(i(e,n[s]));else if(o(e,n[s],r))return n[s]}}function s(e,n,t){if(a(e,n))return n;const o=n.getChildren(t),i=r(e,o,o.length,t);return i&&s(e,i,t)}function a(e,n){return e.isToken(n)&&!i(e,n)}Object.defineProperty(n,"__esModule",{value:!0}),n.isInString=function(e,n,t){const o=function(e,n,t){let o=n;e:for(;;){for(const i of o.getChildren(n)){if(i.getFullStart()>t)break;const n=i.getEnd();if(t<n||t===n&&i.kind===e.SyntaxKind.EndOfFileToken){o=i;continue e}}return o}}(e,n,t);if(o&&(e.isStringLiteral(o)||e.isTemplateLiteralToken(o))){const e=o.getStart(n),i=o.getEnd();if(e<t&&t<i)return!0;if(t===i)return!!o.isUnterminated}},n.findPrecedingToken=function(e,n,t,c,u){return function c(l){if(a(e,l)&&l.kind!==e.SyntaxKind.EndOfFileToken)return l;const f=l.getChildren(t);for(let a=0;a<f.length;a++){const l=f[a];if(n<l.end){if(l.getStart(t,!u)>=n||!o(e,l,t)||i(e,l)){const n=r(e,f,a,t);return n&&s(e,n,t)}return c(l)}}const d=r(e,f,f.length,t);return d&&s(e,d,t)}(c||t)},n.isKeyword=function(e,n){return e.SyntaxKind.FirstKeyword<=n&&n<=e.SyntaxKind.LastKeyword}}]);
// SIG // Begin signature block
// SIG // MIIoNwYJKoZIhvcNAQcCoIIoKDCCKCQCAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // +cHhA2ynpKhWoa0kUGWg1TEgQWsQhSzUOKWVhitSJsGg
// SIG // gg2FMIIGAzCCA+ugAwIBAgITMwAABAO91ZVdDzsYrQAA
// SIG // AAAEAzANBgkqhkiG9w0BAQsFADB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExMB4XDTI0MDkxMjIwMTExM1oX
// SIG // DTI1MDkxMTIwMTExM1owdDELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEeMBwGA1UEAxMVTWljcm9zb2Z0IENvcnBvcmF0aW9u
// SIG // MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA
// SIG // n3RnXcCDp20WFMoNNzt4s9fV12T5roRJlv+bshDfvJoM
// SIG // ZfhyRnixgUfGAbrRlS1St/EcXFXD2MhRkF3CnMYIoeMO
// SIG // MuMyYtxr2sC2B5bDRMUMM/r9I4GP2nowUthCWKFIS1RP
// SIG // lM0YoVfKKMaH7bJii29sW+waBUulAKN2c+Gn5znaiOxR
// SIG // qIu4OL8f9DCHYpME5+Teek3SL95sH5GQhZq7CqTdM0fB
// SIG // w/FmLLx98SpBu7v8XapoTz6jJpyNozhcP/59mi/Fu4tT
// SIG // 2rI2vD50Vx/0GlR9DNZ2py/iyPU7DG/3p1n1zluuRp3u
// SIG // XKjDfVKH7xDbXcMBJid22a3CPbuC2QJLowIDAQABo4IB
// SIG // gjCCAX4wHwYDVR0lBBgwFgYKKwYBBAGCN0wIAQYIKwYB
// SIG // BQUHAwMwHQYDVR0OBBYEFOpuKgJKc+OuNYitoqxfHlrE
// SIG // gXAZMFQGA1UdEQRNMEukSTBHMS0wKwYDVQQLEyRNaWNy
// SIG // b3NvZnQgSXJlbGFuZCBPcGVyYXRpb25zIExpbWl0ZWQx
// SIG // FjAUBgNVBAUTDTIzMDAxMis1MDI5MjYwHwYDVR0jBBgw
// SIG // FoAUSG5k5VAF04KqFzc3IrVtqMp1ApUwVAYDVR0fBE0w
// SIG // SzBJoEegRYZDaHR0cDovL3d3dy5taWNyb3NvZnQuY29t
// SIG // L3BraW9wcy9jcmwvTWljQ29kU2lnUENBMjAxMV8yMDEx
// SIG // LTA3LTA4LmNybDBhBggrBgEFBQcBAQRVMFMwUQYIKwYB
// SIG // BQUHMAKGRWh0dHA6Ly93d3cubWljcm9zb2Z0LmNvbS9w
// SIG // a2lvcHMvY2VydHMvTWljQ29kU2lnUENBMjAxMV8yMDEx
// SIG // LTA3LTA4LmNydDAMBgNVHRMBAf8EAjAAMA0GCSqGSIb3
// SIG // DQEBCwUAA4ICAQBRaP+hOC1+dSKhbqCr1LIvNEMrRiOQ
// SIG // EkPc7D6QWtM+/IbrYiXesNeeCZHCMf3+6xASuDYQ+AyB
// SIG // TX0YlXSOxGnBLOzgEukBxezbfnhUTTk7YB2/TxMUcuBC
// SIG // P45zMM0CVTaJE8btloB6/3wbFrOhvQHCILx41jTd6kUq
// SIG // 4bIBHah3NG0Q1H/FCCwHRGTjAbyiwq5n/pCTxLz5XYCu
// SIG // 4RTvy/ZJnFXuuwZynowyju90muegCToTOwpHgE6yRcTv
// SIG // Ri16LKCr68Ab8p8QINfFvqWoEwJCXn853rlkpp4k7qzw
// SIG // lBNiZ71uw2pbzjQzrRtNbCFQAfmoTtsHFD2tmZvQIg1Q
// SIG // VkzM/V1KCjHL54ItqKm7Ay4WyvqWK0VIEaTbdMtbMWbF
// SIG // zq2hkRfJTNnFr7RJFeVC/k0DNaab+bpwx5FvCUvkJ3z2
// SIG // wfHWVUckZjEOGmP7cecefrF+rHpif/xW4nJUjMUiPsyD
// SIG // btY2Hq3VMLgovj+qe0pkJgpYQzPukPm7RNhbabFNFvq+
// SIG // kXWBX/z/pyuo9qLZfTb697Vi7vll5s/DBjPtfMpyfpWG
// SIG // 0phVnAI+0mM4gH09LCMJUERZMgu9bbCGVIQR7cT5YhlL
// SIG // t+tpSDtC6XtAzq4PJbKZxFjpB5wk+SRJ1gm87olbfEV9
// SIG // SFdO7iL3jWbjgVi1Qs1iYxBmvh4WhLWr48uouzCCB3ow
// SIG // ggVioAMCAQICCmEOkNIAAAAAAAMwDQYJKoZIhvcNAQEL
// SIG // BQAwgYgxCzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNo
// SIG // aW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQK
// SIG // ExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xMjAwBgNVBAMT
// SIG // KU1pY3Jvc29mdCBSb290IENlcnRpZmljYXRlIEF1dGhv
// SIG // cml0eSAyMDExMB4XDTExMDcwODIwNTkwOVoXDTI2MDcw
// SIG // ODIxMDkwOVowfjELMAkGA1UEBhMCVVMxEzARBgNVBAgT
// SIG // Cldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAc
// SIG // BgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEoMCYG
// SIG // A1UEAxMfTWljcm9zb2Z0IENvZGUgU2lnbmluZyBQQ0Eg
// SIG // MjAxMTCCAiIwDQYJKoZIhvcNAQEBBQADggIPADCCAgoC
// SIG // ggIBAKvw+nIQHC6t2G6qghBNNLrytlghn0IbKmvpWlCq
// SIG // uAY4GgRJun/DDB7dN2vGEtgL8DjCmQawyDnVARQxQtOJ
// SIG // DXlkh36UYCRsr55JnOloXtLfm1OyCizDr9mpK656Ca/X
// SIG // llnKYBoF6WZ26DJSJhIv56sIUM+zRLdd2MQuA3WraPPL
// SIG // bfM6XKEW9Ea64DhkrG5kNXimoGMPLdNAk/jj3gcN1Vx5
// SIG // pUkp5w2+oBN3vpQ97/vjK1oQH01WKKJ6cuASOrdJXtjt
// SIG // 7UORg9l7snuGG9k+sYxd6IlPhBryoS9Z5JA7La4zWMW3
// SIG // Pv4y07MDPbGyr5I4ftKdgCz1TlaRITUlwzluZH9TupwP
// SIG // rRkjhMv0ugOGjfdf8NBSv4yUh7zAIXQlXxgotswnKDgl
// SIG // mDlKNs98sZKuHCOnqWbsYR9q4ShJnV+I4iVd0yFLPlLE
// SIG // tVc/JAPw0XpbL9Uj43BdD1FGd7P4AOG8rAKCX9vAFbO9
// SIG // G9RVS+c5oQ/pI0m8GLhEfEXkwcNyeuBy5yTfv0aZxe/C
// SIG // HFfbg43sTUkwp6uO3+xbn6/83bBm4sGXgXvt1u1L50kp
// SIG // pxMopqd9Z4DmimJ4X7IvhNdXnFy/dygo8e1twyiPLI9A
// SIG // N0/B4YVEicQJTMXUpUMvdJX3bvh4IFgsE11glZo+TzOE
// SIG // 2rCIF96eTvSWsLxGoGyY0uDWiIwLAgMBAAGjggHtMIIB
// SIG // 6TAQBgkrBgEEAYI3FQEEAwIBADAdBgNVHQ4EFgQUSG5k
// SIG // 5VAF04KqFzc3IrVtqMp1ApUwGQYJKwYBBAGCNxQCBAwe
// SIG // CgBTAHUAYgBDAEEwCwYDVR0PBAQDAgGGMA8GA1UdEwEB
// SIG // /wQFMAMBAf8wHwYDVR0jBBgwFoAUci06AjGQQ7kUBU7h
// SIG // 6qfHMdEjiTQwWgYDVR0fBFMwUTBPoE2gS4ZJaHR0cDov
// SIG // L2NybC5taWNyb3NvZnQuY29tL3BraS9jcmwvcHJvZHVj
// SIG // dHMvTWljUm9vQ2VyQXV0MjAxMV8yMDExXzAzXzIyLmNy
// SIG // bDBeBggrBgEFBQcBAQRSMFAwTgYIKwYBBQUHMAKGQmh0
// SIG // dHA6Ly93d3cubWljcm9zb2Z0LmNvbS9wa2kvY2VydHMv
// SIG // TWljUm9vQ2VyQXV0MjAxMV8yMDExXzAzXzIyLmNydDCB
// SIG // nwYDVR0gBIGXMIGUMIGRBgkrBgEEAYI3LgMwgYMwPwYI
// SIG // KwYBBQUHAgEWM2h0dHA6Ly93d3cubWljcm9zb2Z0LmNv
// SIG // bS9wa2lvcHMvZG9jcy9wcmltYXJ5Y3BzLmh0bTBABggr
// SIG // BgEFBQcCAjA0HjIgHQBMAGUAZwBhAGwAXwBwAG8AbABp
// SIG // AGMAeQBfAHMAdABhAHQAZQBtAGUAbgB0AC4gHTANBgkq
// SIG // hkiG9w0BAQsFAAOCAgEAZ/KGpZjgVHkaLtPYdGcimwuW
// SIG // EeFjkplCln3SeQyQwWVfLiw++MNy0W2D/r4/6ArKO79H
// SIG // qaPzadtjvyI1pZddZYSQfYtGUFXYDJJ80hpLHPM8QotS
// SIG // 0LD9a+M+By4pm+Y9G6XUtR13lDni6WTJRD14eiPzE32m
// SIG // kHSDjfTLJgJGKsKKELukqQUMm+1o+mgulaAqPyprWElj
// SIG // HwlpblqYluSD9MCP80Yr3vw70L01724lruWvJ+3Q3fMO
// SIG // r5kol5hNDj0L8giJ1h/DMhji8MUtzluetEk5CsYKwsat
// SIG // ruWy2dsViFFFWDgycScaf7H0J/jeLDogaZiyWYlobm+n
// SIG // t3TDQAUGpgEqKD6CPxNNZgvAs0314Y9/HG8VfUWnduVA
// SIG // KmWjw11SYobDHWM2l4bf2vP48hahmifhzaWX0O5dY0Hj
// SIG // Wwechz4GdwbRBrF1HxS+YWG18NzGGwS+30HHDiju3mUv
// SIG // 7Jf2oVyW2ADWoUa9WfOXpQlLSBCZgB/QACnFsZulP0V3
// SIG // HjXG0qKin3p6IvpIlR+r+0cjgPWe+L9rt0uX4ut1eBrs
// SIG // 6jeZeRhL/9azI2h15q/6/IvrC4DqaTuv/DDtBEyO3991
// SIG // bWORPdGdVk5Pv4BXIqF4ETIheu9BCrE/+6jMpF3BoYib
// SIG // V3FWTkhFwELJm3ZbCoBIa/15n8G9bW1qyVJzEw16UM0x
// SIG // ghoKMIIaBgIBATCBlTB+MQswCQYDVQQGEwJVUzETMBEG
// SIG // A1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9u
// SIG // ZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBvcmF0aW9u
// SIG // MSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBTaWduaW5n
// SIG // IFBDQSAyMDExAhMzAAAEA73VlV0POxitAAAAAAQDMA0G
// SIG // CWCGSAFlAwQCAQUAoIGuMBkGCSqGSIb3DQEJAzEMBgor
// SIG // BgEEAYI3AgEEMBwGCisGAQQBgjcCAQsxDjAMBgorBgEE
// SIG // AYI3AgEVMC8GCSqGSIb3DQEJBDEiBCB7LPuu6707RD8/
// SIG // THr92I9uppbu4V9umTExjhFO2aS5lzBCBgorBgEEAYI3
// SIG // AgEMMTQwMqAUgBIATQBpAGMAcgBvAHMAbwBmAHShGoAY
// SIG // aHR0cDovL3d3dy5taWNyb3NvZnQuY29tMA0GCSqGSIb3
// SIG // DQEBAQUABIIBABp0R+9Ezhtdc3L7Ijvv5FBySvk3xmzZ
// SIG // fdQnW0So8U/D0sa67T0m/MGUzUQy2Yf2EyLXyy7rP8F6
// SIG // mo0tPRj0PLvC5Q85kj7ZG5TZm2H7jaRKS6rjaP5IU+ZT
// SIG // cS8bQpzCaeazPrOTEaEHO8bAMtR8xpm7HE7LTxmWMtlQ
// SIG // i+D96PVFvN+gdSE0cEzUriW0YeUOthBKExowVVf0wsb6
// SIG // aJwg/fq6beRzknDp+vyrI0FiTz8KLByDcY7sFOKhBxrP
// SIG // +seS1gTdk9Zvj0xfFUK+VrBC5e0ohRzuq6chPDweLDvE
// SIG // 0btAOt6EIWwhYtfSt3Uns48LpMtWKNg2l8VP+JdPAVSr
// SIG // 85qhgheUMIIXkAYKKwYBBAGCNwMDATGCF4Awghd8Bgkq
// SIG // hkiG9w0BBwKgghdtMIIXaQIBAzEPMA0GCWCGSAFlAwQC
// SIG // AQUAMIIBUgYLKoZIhvcNAQkQAQSgggFBBIIBPTCCATkC
// SIG // AQEGCisGAQQBhFkKAwEwMTANBglghkgBZQMEAgEFAAQg
// SIG // ARY522HSD/f3Zug0uvR0FpvjfMiFRxB+eciloLJYOn0C
// SIG // BmgSpjzu/hgTMjAyNTA1MDIxOTE1MDguNDE4WjAEgAIB
// SIG // 9KCB0aSBzjCByzELMAkGA1UEBhMCVVMxEzARBgNVBAgT
// SIG // Cldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAc
// SIG // BgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjElMCMG
// SIG // A1UECxMcTWljcm9zb2Z0IEFtZXJpY2EgT3BlcmF0aW9u
// SIG // czEnMCUGA1UECxMeblNoaWVsZCBUU1MgRVNOOjhEMDAt
// SIG // MDVFMC1EOTQ3MSUwIwYDVQQDExxNaWNyb3NvZnQgVGlt
// SIG // ZS1TdGFtcCBTZXJ2aWNloIIR6jCCByAwggUIoAMCAQIC
// SIG // EzMAAAINDXe+ezaPf+MAAQAAAg0wDQYJKoZIhvcNAQEL
// SIG // BQAwfDELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hp
// SIG // bmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoT
// SIG // FU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEmMCQGA1UEAxMd
// SIG // TWljcm9zb2Z0IFRpbWUtU3RhbXAgUENBIDIwMTAwHhcN
// SIG // MjUwMTMwMTk0MzAxWhcNMjYwNDIyMTk0MzAxWjCByzEL
// SIG // MAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hpbmd0b24x
// SIG // EDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jv
// SIG // c29mdCBDb3Jwb3JhdGlvbjElMCMGA1UECxMcTWljcm9z
// SIG // b2Z0IEFtZXJpY2EgT3BlcmF0aW9uczEnMCUGA1UECxMe
// SIG // blNoaWVsZCBUU1MgRVNOOjhEMDAtMDVFMC1EOTQ3MSUw
// SIG // IwYDVQQDExxNaWNyb3NvZnQgVGltZS1TdGFtcCBTZXJ2
// SIG // aWNlMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKC
// SIG // AgEAsX+oB93x98IJhTeU1fM0RfzdEyYLllGAq0GT1JsA
// SIG // PkIUskjae5IhphWXeG+c7AHzdFofj8yCb2LFTIFQmmXG
// SIG // 1a/wyQX3scJ0H3Pn4CIzogM2Eo4OxLUjnVqC5PNW/5ho
// SIG // hiAQvvny3EgM22TtyrVuD0SjkrHQ/+XkXbTXGvxHeBAH
// SIG // FE99swBkvUqv/+JA0/dihM0TVyY/M9zQy+KBhswkDLwP
// SIG // N7UBuZq53qcRdDrDXyih2BzKo8jhD+aU4Cn0+VzH89LB
// SIG // BugEQI5Vye/Uhv46Hrh4fE6gvnCTK8vcAe2S63TABcvK
// SIG // xzd1tfpJKcOOyqPvYs3tm5Xvj+YzGqrnytN9wbvthV1U
// SIG // iqYYpdo9KgVFqZ+bwbX9lvoLFasnQPnrHgR32MXA2ePu
// SIG // aEaMgwINLVvX+/bzgL/dT+G08xvgWTwWwzM0g8Jo/AEi
// SIG // g5VcK7k3hdXpuzVonnVgAftvYBR/3PzTdbk0C78BgyDs
// SIG // NQ3C5wg9FVzS+5jr4n2LzVroEwB3DBS1ZQNY3jh9+RPy
// SIG // FvanCiqq2oey8iWsf+jOVB7MELoBoC64ojDRL89yKlvg
// SIG // IiCdPe4mMzBs7A6my0lM7SIwxLfcXyBx1LP6d4/QZK9m
// SIG // zz9b+M/L6f+OQpN5nxW1thdBLYzgpxjzKyTY4xoRBvcw
// SIG // CtddOQYNpCe0Z/Fk1juSGw/6w7UCAwEAAaOCAUkwggFF
// SIG // MB0GA1UdDgQWBBQVl46Q5g4Da07CEo9uko1Nf2DDgDAf
// SIG // BgNVHSMEGDAWgBSfpxVdAF5iXYP05dJlpxtTNRnpcjBf
// SIG // BgNVHR8EWDBWMFSgUqBQhk5odHRwOi8vd3d3Lm1pY3Jv
// SIG // c29mdC5jb20vcGtpb3BzL2NybC9NaWNyb3NvZnQlMjBU
// SIG // aW1lLVN0YW1wJTIwUENBJTIwMjAxMCgxKS5jcmwwbAYI
// SIG // KwYBBQUHAQEEYDBeMFwGCCsGAQUFBzAChlBodHRwOi8v
// SIG // d3d3Lm1pY3Jvc29mdC5jb20vcGtpb3BzL2NlcnRzL01p
// SIG // Y3Jvc29mdCUyMFRpbWUtU3RhbXAlMjBQQ0ElMjAyMDEw
// SIG // KDEpLmNydDAMBgNVHRMBAf8EAjAAMBYGA1UdJQEB/wQM
// SIG // MAoGCCsGAQUFBwMIMA4GA1UdDwEB/wQEAwIHgDANBgkq
// SIG // hkiG9w0BAQsFAAOCAgEAHB/IE+NsFUPBkGfjByFzSfne
// SIG // HUJ9PrzxvwHRb08ZP7bzjpANou4Z6BDhBXEm80uRHSFQ
// SIG // jIuWORIlk/IRCZM7XXquYV7VFrD2PtpWElPR3YSg30Q3
// SIG // 48FWfZHwrYay0Qt48PRWoeTABiPTB0hE6Pa+/nLsfAFK
// SIG // kN4ZmjNo5kzVr6xTi2ztNQdXIjRm7xNsSvq28wQYf7ln
// SIG // d5N9KWrui6dGs4O6w2BjBr/I1qLSC36SjAXgkKLnzLCP
// SIG // 4M8aOPpPXamAZAMDJO4dYX9t8nG+Vp4EjEFf1MZd0sW4
// SIG // PuuCiuhgrE0hus+8zU4Gy+vTzBn8FdnJ2sGPeSMtnICe
// SIG // nGzDkUVHGM77LlHfKBEPO3FuF6qgS64XnsCY0Tk+mMlq
// SIG // 0NqEPo8xitliaBjZ32p5jX6kJ1SqYt9tKJ2oPeSzBY4/
// SIG // AcK35WLDvZBDosuyOVJ4DwZhfkR0gr9oRUq/Jyu4Lbos
// SIG // vUlDij+hW0Ds8jEbyd1MbJqx9/OO+18xE2/7WSwI9dE3
// SIG // WBX5I5BXhY5W94jE0rP1ayUhdK9Lzjf6MWaigmPFmXs/
// SIG // rkD9E68tEEPcEaeQ9hWCfBNbiF/y78LhZAXVn+1+6pW/
// SIG // qc3bz6UgHGgoASFk5TJ9m1oJFObj88UTx4f9LvhSO/3k
// SIG // QmRmyzNqp598l0Lzv4fcEJvKSAAUQL6HWoldG1Zh2xEw
// SIG // ggdxMIIFWaADAgECAhMzAAAAFcXna54Cm0mZAAAAAAAV
// SIG // MA0GCSqGSIb3DQEBCwUAMIGIMQswCQYDVQQGEwJVUzET
// SIG // MBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVk
// SIG // bW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBvcmF0
// SIG // aW9uMTIwMAYDVQQDEylNaWNyb3NvZnQgUm9vdCBDZXJ0
// SIG // aWZpY2F0ZSBBdXRob3JpdHkgMjAxMDAeFw0yMTA5MzAx
// SIG // ODIyMjVaFw0zMDA5MzAxODMyMjVaMHwxCzAJBgNVBAYT
// SIG // AlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQH
// SIG // EwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29y
// SIG // cG9yYXRpb24xJjAkBgNVBAMTHU1pY3Jvc29mdCBUaW1l
// SIG // LVN0YW1wIFBDQSAyMDEwMIICIjANBgkqhkiG9w0BAQEF
// SIG // AAOCAg8AMIICCgKCAgEA5OGmTOe0ciELeaLL1yR5vQ7V
// SIG // gtP97pwHB9KpbE51yMo1V/YBf2xK4OK9uT4XYDP/XE/H
// SIG // ZveVU3Fa4n5KWv64NmeFRiMMtY0Tz3cywBAY6GB9alKD
// SIG // RLemjkZrBxTzxXb1hlDcwUTIcVxRMTegCjhuje3XD9gm
// SIG // U3w5YQJ6xKr9cmmvHaus9ja+NSZk2pg7uhp7M62AW36M
// SIG // EBydUv626GIl3GoPz130/o5Tz9bshVZN7928jaTjkY+y
// SIG // OSxRnOlwaQ3KNi1wjjHINSi947SHJMPgyY9+tVSP3PoF
// SIG // VZhtaDuaRr3tpK56KTesy+uDRedGbsoy1cCGMFxPLOJi
// SIG // ss254o2I5JasAUq7vnGpF1tnYN74kpEeHT39IM9zfUGa
// SIG // RnXNxF803RKJ1v2lIH1+/NmeRd+2ci/bfV+Autuqfjbs
// SIG // Nkz2K26oElHovwUDo9Fzpk03dJQcNIIP8BDyt0cY7afo
// SIG // mXw/TNuvXsLz1dhzPUNOwTM5TI4CvEJoLhDqhFFG4tG9
// SIG // ahhaYQFzymeiXtcodgLiMxhy16cg8ML6EgrXY28MyTZk
// SIG // i1ugpoMhXV8wdJGUlNi5UPkLiWHzNgY1GIRH29wb0f2y
// SIG // 1BzFa/ZcUlFdEtsluq9QBXpsxREdcu+N+VLEhReTwDwV
// SIG // 2xo3xwgVGD94q0W29R6HXtqPnhZyacaue7e3PmriLq0C
// SIG // AwEAAaOCAd0wggHZMBIGCSsGAQQBgjcVAQQFAgMBAAEw
// SIG // IwYJKwYBBAGCNxUCBBYEFCqnUv5kxJq+gpE8RjUpzxD/
// SIG // LwTuMB0GA1UdDgQWBBSfpxVdAF5iXYP05dJlpxtTNRnp
// SIG // cjBcBgNVHSAEVTBTMFEGDCsGAQQBgjdMg30BATBBMD8G
// SIG // CCsGAQUFBwIBFjNodHRwOi8vd3d3Lm1pY3Jvc29mdC5j
// SIG // b20vcGtpb3BzL0RvY3MvUmVwb3NpdG9yeS5odG0wEwYD
// SIG // VR0lBAwwCgYIKwYBBQUHAwgwGQYJKwYBBAGCNxQCBAwe
// SIG // CgBTAHUAYgBDAEEwCwYDVR0PBAQDAgGGMA8GA1UdEwEB
// SIG // /wQFMAMBAf8wHwYDVR0jBBgwFoAU1fZWy4/oolxiaNE9
// SIG // lJBb186aGMQwVgYDVR0fBE8wTTBLoEmgR4ZFaHR0cDov
// SIG // L2NybC5taWNyb3NvZnQuY29tL3BraS9jcmwvcHJvZHVj
// SIG // dHMvTWljUm9vQ2VyQXV0XzIwMTAtMDYtMjMuY3JsMFoG
// SIG // CCsGAQUFBwEBBE4wTDBKBggrBgEFBQcwAoY+aHR0cDov
// SIG // L3d3dy5taWNyb3NvZnQuY29tL3BraS9jZXJ0cy9NaWNS
// SIG // b29DZXJBdXRfMjAxMC0wNi0yMy5jcnQwDQYJKoZIhvcN
// SIG // AQELBQADggIBAJ1VffwqreEsH2cBMSRb4Z5yS/ypb+pc
// SIG // FLY+TkdkeLEGk5c9MTO1OdfCcTY/2mRsfNB1OW27DzHk
// SIG // wo/7bNGhlBgi7ulmZzpTTd2YurYeeNg2LpypglYAA7AF
// SIG // vonoaeC6Ce5732pvvinLbtg/SHUB2RjebYIM9W0jVOR4
// SIG // U3UkV7ndn/OOPcbzaN9l9qRWqveVtihVJ9AkvUCgvxm2
// SIG // EhIRXT0n4ECWOKz3+SmJw7wXsFSFQrP8DJ6LGYnn8Atq
// SIG // gcKBGUIZUnWKNsIdw2FzLixre24/LAl4FOmRsqlb30mj
// SIG // dAy87JGA0j3mSj5mO0+7hvoyGtmW9I/2kQH2zsZ0/fZM
// SIG // cm8Qq3UwxTSwethQ/gpY3UA8x1RtnWN0SCyxTkctwRQE
// SIG // cb9k+SS+c23Kjgm9swFXSVRk2XPXfx5bRAGOWhmRaw2f
// SIG // pCjcZxkoJLo4S5pu+yFUa2pFEUep8beuyOiJXk+d0tBM
// SIG // drVXVAmxaQFEfnyhYWxz/gq77EFmPWn9y8FBSX5+k77L
// SIG // +DvktxW/tM4+pTFRhLy/AsGConsXHRWJjXD+57XQKBqJ
// SIG // C4822rpM+Zv/Cuk0+CQ1ZyvgDbjmjJnW4SLq8CdCPSWU
// SIG // 5nR0W2rRnj7tfqAxM328y+l7vzhwRNGQ8cirOoo6CGJ/
// SIG // 2XBjU02N7oJtpQUQwXEGahC0HVUzWLOhcGbyoYIDTTCC
// SIG // AjUCAQEwgfmhgdGkgc4wgcsxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xJTAjBgNVBAsTHE1pY3Jvc29mdCBBbWVyaWNhIE9w
// SIG // ZXJhdGlvbnMxJzAlBgNVBAsTHm5TaGllbGQgVFNTIEVT
// SIG // Tjo4RDAwLTA1RTAtRDk0NzElMCMGA1UEAxMcTWljcm9z
// SIG // b2Z0IFRpbWUtU3RhbXAgU2VydmljZaIjCgEBMAcGBSsO
// SIG // AwIaAxUAeywsKI+ht50OzsYZf6X53uMgEz2ggYMwgYCk
// SIG // fjB8MQswCQYDVQQGEwJVUzETMBEGA1UECBMKV2FzaGlu
// SIG // Z3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwGA1UEChMV
// SIG // TWljcm9zb2Z0IENvcnBvcmF0aW9uMSYwJAYDVQQDEx1N
// SIG // aWNyb3NvZnQgVGltZS1TdGFtcCBQQ0EgMjAxMDANBgkq
// SIG // hkiG9w0BAQsFAAIFAOu/HtMwIhgPMjAyNTA1MDIxMDM3
// SIG // MDdaGA8yMDI1MDUwMzEwMzcwN1owdDA6BgorBgEEAYRZ
// SIG // CgQBMSwwKjAKAgUA678e0wIBADAHAgEAAgIx9jAHAgEA
// SIG // AgITfTAKAgUA68BwUwIBADA2BgorBgEEAYRZCgQCMSgw
// SIG // JjAMBgorBgEEAYRZCgMCoAowCAIBAAIDB6EgoQowCAIB
// SIG // AAIDAYagMA0GCSqGSIb3DQEBCwUAA4IBAQBKS/TZ6XIE
// SIG // gFZ1MxbfRNjQCrLJG7Zi4PiHSdjKhI1ctgfXE9kcObL6
// SIG // EcEX3ujYHIUKWMrQVmbElfQio7HwKvMjNz9cPf5O9P/S
// SIG // YfBP9RH0yuR0VUsQqej/Q7bOfSdhbP91TxRo8aDEAbVm
// SIG // YCwDoSqsqykyzGr7zNJrKHfCyhYsjELo7XlOaprdWKjr
// SIG // gG2oMOLcfdqhYryRkgmWgBZHetsSzucqQw0Fp5tIXeXr
// SIG // 58lxJaEiqjN9dy7jH8u2YPH7Ds6KUYO656WjyKlosv4H
// SIG // BPphieHwa3nC3x+YSggbF6n/AgIAwolcyZRhqShs3mtK
// SIG // iiaUzw/G1le4ZTjqir+FP7GbMYIEDTCCBAkCAQEwgZMw
// SIG // fDELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hpbmd0
// SIG // b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1p
// SIG // Y3Jvc29mdCBDb3Jwb3JhdGlvbjEmMCQGA1UEAxMdTWlj
// SIG // cm9zb2Z0IFRpbWUtU3RhbXAgUENBIDIwMTACEzMAAAIN
// SIG // DXe+ezaPf+MAAQAAAg0wDQYJYIZIAWUDBAIBBQCgggFK
// SIG // MBoGCSqGSIb3DQEJAzENBgsqhkiG9w0BCRABBDAvBgkq
// SIG // hkiG9w0BCQQxIgQgDQpdBcU81SVXt7yztLUUF8jjyqlT
// SIG // itTkNX+1qHQC1zAwgfoGCyqGSIb3DQEJEAIvMYHqMIHn
// SIG // MIHkMIG9BCBj6geU7CRqiPDrpJwSrUr29BX6usOTUrOb
// SIG // 3k1cJGv4QDCBmDCBgKR+MHwxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xJjAkBgNVBAMTHU1pY3Jvc29mdCBUaW1lLVN0YW1w
// SIG // IFBDQSAyMDEwAhMzAAACDQ13vns2j3/jAAEAAAINMCIE
// SIG // IMItyTIzLnMqyM6DiackyGUNFAziLU9IgIh6iur4QRPL
// SIG // MA0GCSqGSIb3DQEBCwUABIICAArDvuTIeO3gs6BPRl9u
// SIG // otVPlWD8LgqWUlJhITFEwz/Yw4uTEhLOsJdRBkq2hpRg
// SIG // AXRuptCxIqr/mroPQ4jVZopNpUqnd5f2TEkvDpT193y1
// SIG // AqA9orF92lSzHLzBWW0xnj6P8n+zAY10PAE93pGcVAGn
// SIG // Y1lsV6iw+iI1yjQtHmuVFtohke52nlxc450v4FxiLQcu
// SIG // 2XVqrctNuAQB2x3Zef7m/v14J4NXr/b9vSFJqoVh8Qar
// SIG // vI8Zf5ZvQRPr6RoopsznW4grRaqsIjTiVXiuADVVqchl
// SIG // lYcLIoH5AObLzSd3HMpa5+2RYrJnTCVyI8r9igcZIPuQ
// SIG // bHpN+kDYxocl+Ws0swhsUIVyhd2Ig3Cp9VNnDsEbGmN2
// SIG // Ux5Y5LbA+ykzeapDcetnau36YUuw3b1dl1PIqzrOyrLy
// SIG // 4ddVEVk+26dZzVlXqYD6kAoZl8zJdFxidlmaBG0vymeX
// SIG // ZEOD/yNzyV5IvWGclNzrv0FTYZEdvmAYZNBJu7EVKDAf
// SIG // fbqtnq5FuhhJrI7ebowLpucJ+7C8QpC02KbQkojGG0/F
// SIG // EAsiQ+wZZuzbLfF5kGf1w+I7d1ErNWsEp+qzdUR5DbE5
// SIG // mrTeuK0TmfFp2NemlY9eKXbWv6GHD2EpZDesVNp10m4c
// SIG // 9X8BL1THhefOIA4oFpDpw79DnHh1gPy7uxz8pkGxgg28
// SIG // gr3x
// SIG // End signature block
